mod/dagmodifier.go

```diff
@@ -627,3 +647,315 @@ func (dm *DagModifier) GetCtx() context.Context {
 func (dm *DagModifier) GetDserv() ipld.DAGService {
 	return dm.dagserv
 }
+
+func (mdm *MetaDagModifier) GetDb() *help.DagBuilderHelper {
+	return mdm.db
+}
+
+// AddMetadata puts the given `metadata` items to the given `root`
+// that is the top node of a BTFS file DAG and returns a new root for the BTFS file DAG.
+//
+// There are three scenarios possible:
+// Scenario #1. No existing metadata for the `root`. So this scenario is
+//    to create a new metadata sub-DAG.
+// Scenario #2. There are existing keys for the given `metadata` items.
+//    Then this is to update the metadata sub-DAG with the given items.
+// Scenario #3. There are no pre-existing metadata items for the given `metadata` items.
+//    Then this scenario is to append given items to the metadata sub-DAG.
+// TODO: trickle format
+// Preconditions include that first the given `metadata` is in compact JSON format.
+func (mdm *MetaDagModifier) AddMetadata(root ipld.Node, metadata []byte) (ipld.Node, error) {
+	// Read the existing metadata map.
+	b, err := mdm.readMetadataBytes(root)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine the specific scenario.
+	// Scenario #1:
+	var newMetaNode ipld.Node
+	var children *ft.DagMetaNodes
+	if b == nil {
+		// TODO: Get SuperMeta items and add to metadata.
+		// Create a metadata sub-DAG
+		newMetaNode, err = mdm.buildNewMetaDataDag(metadata)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		children, err := ft.GetChildrenForDagWithMeta(mdm.ctx, root, mdm.dagserv)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create existing map and input metadata map and check.
+		m := make(map[string]interface{})
+		err = json.Unmarshal(b, &m)
+		if err != nil {
+			return nil, err
+		}
+
+		inputM := make(map[string]interface{})
+		err = json.Unmarshal(metadata, &inputM)
+		if err != nil {
+			return nil, err
+		}
+
+		exists := intersects(m, inputM)
+		// Scenario #2: Update case.
+		if exists {
+			// iterate the inputM to put its (k, v) pairs to existing map.
+			for k, v := range inputM {
+				m[k] = v
```

Comment:
But this may insert new keys into `m` as well, shouldn't this need `mdm.WriteAt()` in order to append?
Related question is, does `buildNewMetadataDag` handle the "append more bytes" automatically?

Reply:
This is update case.

Reply:
Will add Truncate() per talk